### ISSUE TITLE
Probleme with line ODT generation

### ISSUE THREAD

**Issue by Ninjabil** [Dolibarr/dolibarr#16184]
# Instructions
*This is a template to help you report good issues. You may use [Github Markdown](https://help.github.com/articles/getting-started-with-writing-and-formatting-on-github/) syntax to format your issue report.*
*Please:*
- *replace the bracket enclosed texts with meaningful information*
- *remove any unused sub-section*

# Bug
Generation invoices ODT

## Environment
- Dolibarr 13: 13
- Debian 
- Nginx
- 7.4
- Maria DB

## Expected and actual behavior
It generates line with html balises like this 
![Capture](https://user-images.githubusercontent.com/78621823/107091490-2c74af80-6802-11eb-9146-8aee66109a3f.PNG)

## Steps to reproduce the behavior
When I generate an odt file

**Comment by ksar-ksar** [comment]
Same problem here : https://www.dolibarr.fr/forum/t/upgrade-en-13-0-0-probleme-daccentues-lors-de-la-generation-de-documents/35641/

**Comment by tootai** [comment]
Hello. I face the same problem. I opened the above case at dolibarr.fr

Daniel

**Comment by tootai** [comment]
I just upgraded to 13.0.1, problem still existing. Any update about this matter ?

**Comment by ddhana** [comment]
Before updating to 13.0.1 I stored the 10.x used before. After checking some PHP sources it seems that the problem is originated in 
...`/includes/odtphp/odf.php`.
WorkAround (qiuck & dirty): Using the sources from 10.x there are no html tags in the produced odt output. Take the complete folder ...`/includes/odtphp` from version 10.x.

In 13.0.1 the function
`public function convertVarToOdf($value, $encode = true, $charset = 'ISO-8859')`
seems to be the source of trouble. The replacements there are not reached and do not transform the necessary tags. I observed that the 
`// Check if the value includes html tags`
fails and falls into the else part. There the two lines would help to suppress the html tags and to preserve the line breaks:
```
$convertedValue=dol_string_nohtmltag($convertedValue, 0);
$convertedValue = preg_replace('/(\r\n|\r|\n)/i', "<text:line-break/>", $convertedValue);
```
But this would also be a dirty workaround. If somebody having more knowledge on ODF converting, he could produce a better solution.


Using 13.0.1 unfortunately most of the ODT output cannot be used. So it should be a high priority bug.
I observed that some PDF output might also be affected.

Thanks a lot for fixing this gamestopper bug !!!

**Comment by ddhana** [comment]
Using 13.0.2 does not solve this issue.

**Comment by tootai** [comment]
Hello. Anyone on this ? Still strunggeling with this bug :(

**Comment by chrkorn** [comment]
This is not a bug with ODT generation. It is the result of FSCKEditor saving HTML encoded texts in the database.
The FSCKEditor module has settings in Dolibarr that allow it to be enabled/disabled for particular areas of Dolibarr, and there is an explicit warning about enabling it for invoices etc. 
For me, disabling FSCKEditor solved the problem, but I had to edit /fix the relevant texts as well.

**Comment by tootai** [comment]
Hi

Le 22/03/2021 à 10:28, Christian Korn a écrit :
>
> This is not a bug with ODT generation. It is the result of FSCKEditor 
> saving HTML encoded texts in the database.
> The FSCKEditor module has settings in Dolibarr that allow it to be 
> enabled/disabled for particular areas of Dolibarr, and there is an 
> explicit warning about enabling it for invoices etc.
> For me, disabling FSCKEditor solved the problem, but I had to edit 
> /fix the relevant texts as well.
>
If true, why does this suddenly appears in 13.0.x then ?

-- 

Daniel Huhardeaux
***@***.***	      ***@***.***
***@***.***		    tootaiNET

**Comment by ddhana** [comment]
My intention would not be to disable FCKEditor. The better solution would be to do any kind of reformatting for the output into other than HTML formats.
Storing text into the database in HTML format is probably also not a good idea because it destoys any possibility to search for text, especially if it contains special characters like ÄÖÜäöüß or any other.
And it is not only related to invoices. It affects all entries going into the database and every kind of documents generated via ODT templates.

**Comment by libremaster** [comment]
**J'ai modifié la fonction comme ceci, à tester :**

```
public function convertVarToOdf($value, $encode = true, $charset = 'ISO-8859')
	{
		$convertedValue = ($charset == 'ISO-8859') ? utf8_encode($value) : $value;

		// Check if the value includes html tags
		if ($this->_hasHtmlTag($value) === true) {
			// Default styles for strong/b, i/em, u, s, sub & sup
			$automaticStyles = array(
				'<style:style style:name="boldText" style:family="text"><style:text-properties fo:font-weight="bold" style:font-weight-asian="bold" style:font-weight-complex="bold" /></style:style>',
				'<style:style style:name="italicText" style:family="text"><style:text-properties fo:font-style="italic" style:font-style-asian="italic" style:font-style-complex="italic" /></style:style>',
				'<style:style style:name="underlineText" style:family="text"><style:text-properties style:text-underline-style="solid" style:text-underline-width="auto" style:text-underline-color="font-color" /></style:style>',
				'<style:style style:name="strikethroughText" style:family="text"><style:text-properties style:text-line-through-style="solid" style:text-line-through-type="single" /></style:style>',
				'<style:style style:name="subText" style:family="text"><style:text-properties style:text-position="sub 58%" /></style:style>',
				'<style:style style:name="supText" style:family="text"><style:text-properties style:text-position="super 58%" /></style:style>'
			);

			$convertedValue = $this->_replaceHtmlWithOdtTag($this->_getDataFromHtml($value), $customStyles, $fontDeclarations);

			foreach ($customStyles as $key => $val) {
				array_push($automaticStyles, '<style:style style:name="customStyle' . $key . '" style:family="text">' . $val . '</style:style>');
			}

			// Join the styles and add them to the content xml
			$styles = '';
			foreach ($automaticStyles as $style) {
				if (strpos($this->contentXml, $style) === false) {
					$styles .= $style;
				}
			}
			$this->contentXml = str_replace('</office:automatic-styles>', $styles . '</office:automatic-styles>', $this->contentXml);

			// Join the font declarations and add them to the content xml
			$fonts = '';
			foreach ($fontDeclarations as $font) {
				if (strpos($this->contentXml, 'style:name="' . $font . '"') === false) {
					$fonts .= '<style:font-face style:name="' . $font . '" svg:font-family="\'' . $font . '\'" />';
				}
			}
			$this->contentXml = str_replace('</office:font-face-decls>', $fonts . '</office:font-face-decls>', $this->contentXml);
		}
		else $convertedValue = preg_replace('/(\r\n|\r|\n)/i', "<text:line-break/>", $value);

		$convertedValue = preg_replace( "/\r|\n/", "", html_entity_decode($convertedValue, ENT_COMPAT, 'UTF-8'));

		return $convertedValue;
	}
```

**Comment by st31ny** [comment]
This unfortunately crashes my ODT→PDF conversion.

Currently, I have this line added as *first* line of `convertVarToOdf()` as a workaround:
```php
$value = html_entity_decode(strip_tags($value));
```
This gives me a working ODT→PDF conversion without any HTML tags.

**Comment by libremaster** [comment]
You can activate the log dolibarr module, loglevel 'notice' and add :

`dol_syslog($convertedValue, LOG_NOTICE);`

before 'return'.
After, you can check dolibarr.log to see the content.

You can also uncompress the ODT with gzip and to see the file content.xml if there is something wrong.

**Comment by st31ny** [comment]
> You can also uncompress the ODT with gzip and to see the file content.xml if there is something wrong.

Ok, I'll give that a try once I find some free time.

**Comment by oas-services** [comment]
I did the same as https://github.com/Dolibarr/dolibarr/issues/16184#issuecomment-879250379 
It work's. Thank's a lot for this workaround.

**Comment by lucky4ever2** [comment]
when will the fix be adopted in the dolibarr code?

**Comment by PROPC-IT** [comment]
Hello

The workaround proposed above here reported below did work well

> Currently, I have this line added as _first_ line of `convertVarToOdf()` as a workaround:
> $value = html_entity_decode(strip_tags($value));

However this does not solve the problem of special characters which are still outputted as code, e.g. instead of L' the output is L&#39;

in the file ...\include\odtphp\odf.php at this point the code is the following:

`public function convertVarToOdf($value, $encode = true, $charset = 'ISO-8859')
	{
		$value = html_entity_decode(strip_tags($value));
		$value = $encode ? htmlspecialchars($value) : $value;
		$value = ($charset == 'ISO-8859') ? utf8_encode($value) : $value;
		$convertedValue = $value;`

Apparently should convert special characters, but does not...
Any idea how to fix this too?

Btw, as a further clue, I noticed that this problem, at least here, appeared after an upgrade, a clean installation of version 14 did not show this behaviour, not at least until I activated FCKeditor module

In fact the culprit is the FCKeditor, does anyone know how to correct this problem?
Thank you

**Comment by playcock** [comment]
@PROPC-IT 
I'm not convinced that FCKeditor is the culprit. I have disabled the editor and typed in html only to have the html tags displayed on the .odt file.

example: 
`<strong>Bold Text</strong>`

**Comment by playcock** [comment]
For me, when the $value gets passed to the convertVarToOdf function, it arrives where $encode = true. This then takes the variable 

`<strong>Bold Text</strong> `
and converts it to
`&lt;strong&gt;Bold Text&lt;/strong&gt;`
on the line 
`$value = $encode ? htmlspecialchars($value) : $value;`

So then the variable fails the following check and doesn't get the styles applied to it when the document gets generated to an .odt file
```
// Check if the value includes html tags
if ($this->_hasHtmlTag($value) === true) {
```

**Comment by playcock** [comment]
I will admit that some of the styles applied by the CKEditor completely disappear, like bulleted lists, when stepping through 
```
// Check if the value includes html tags
if ($this->_hasHtmlTag($value) === true) {
```
But I would be happy with just the automatic styles for 
`// Default styles for strong/b, i/em, u, s, sub & sup`

**Comment by playcock** [comment]
@Ninjabil @PROPC-IT @lucky4ever2 @st31ny @libremaster @ddhana @chrkorn @tootai @ksar-ksar 
The public function convertVarToOdf($value, $encode = true, $charset = 'ISO-8859') odf.php has nothing wrong with it. It's just that every time ->setVars is called, it passes the enocde value as true. At least in doc_generic_odt_modules.php

Example:
`->setVars($key, $val, true, 'UTF-8');`

There is nothing that you can do to pass encode as false when generating ODT documents. We have tried hooks quite extensively and still cannot get the Dolibarr system to bend so that the substitution variables get passwed with $encode = false.

We know it's an issue in Dolibarr and we don't want to modify the core.

**Comment by playcock** [comment]
If the hook functionality would allow us to work with ODT generation and allow us to intervene to set $encode = false and to even add more html tags to covert to ODT, that would be ideal.

**Comment by playcock** [comment]
The issue has also been reported on the Dolibarr Internation Forum: https://www.dolibarr.org/forum/t/dolibarr-v-13-prints-special-characters-in-odt-files/20199

**Comment by hregis** [comment]
@playcock yep and in v15 ?

**Comment by playcock** [comment]
@hregis We are on v 14.0.5. I am unsure about v 15 at the moment.

**Comment by playcock** [comment]
@hregis 
Following the steps in v 15 compared to v 14.0.5 to handle HTML to ODT tags appears the same with setVars $encode being set as true. 

> If the hook functionality would allow us to work with ODT generation and allow us to intervene to set $encode = false and to even add more html tags to covert to ODT, that would be ideal.

**Comment by thomas-Ngr** [comment]
@playcock I created PR #22753 for this issue, if you want to have a look. Your work on it saved me a lot of time investigating, thanks. I hope I did not forget a special case.

**Comment by warnerbryce** [comment]
@thomas-Ngr I tested your PR and it worked for me. Thank you for this work.

**Comment by tootai** [comment]
I agree, still not OK in 16.0.3

**Comment by jccs1953** [comment]
This problem still exists in 16.0.4. My Client is still using 11.0.4 because of this problem. I upgraded a cloned copy of her system to try an troubleshoot the problem but as yet I still get the results as shown below / does anyone have a fix for this when you enter a return in the editor to add the next description line in the same row on the invoice?
![image](https://user-images.githubusercontent.com/17952470/218272861-1cb4e3cf-953b-4068-808f-2eea2a339d6c.png)

**Comment by warnerbryce** [comment]
@jccs1953 @tootai you need to have at least php 7.4 and use the version of Easya-Solutions/dolibarr for the moment this fix isn’t merged into the dolibarr/dolibarr master branch. 
Or you could also reuse the files in this PR to see if it works. 
This PR helped me too 
https://github.com/Dolibarr/dolibarr/pull/23771

**Comment by jccs1953** [comment]
I have tried version 12.0.4 and it works correctly - I believe the problem starts on Version 13.0.1 and versions there after this has been a bug for a long time. I wanted to move my client to a more recent version with the latest enhancements but may need to wait to see if they correct it

**Comment by jccs1953** [comment]
A temporary work around I have found is to use all the files from the folder located at **/includes/odtphp** and copying the folder **“odtphp”** from version 11.0.4 to 16.0.4 by copying the whole folder to the new installation. Rename the folder **odtphp** to **odtphp2** before copying over the old folder. This seems to work

**Comment by jccs1953** [comment]
> This problem still exists in 16.0.4. My Client is still using 11.0.4 because of this problem. I upgraded a cloned copy of her system to try an troubleshoot the problem but as yet I still get the results as shown below / does anyone have a fix for this when you enter a return in the editor to add the next description line in the same row on the invoice? ![image](https://user-images.githubusercontent.com/17952470/218272861-1cb4e3cf-953b-4068-808f-2eea2a339d6c.png)

Problem still exists in 17.0.1

**Comment by Thatoo** [comment]
The problem can't be fixed in v 17.0.1 as the fix (https://github.com/Dolibarr/dolibarr/commit/fa0528a5d76294376d1b95fca1eeec30a21f020d) has apparently only been merged in develop branch 3 weeks ago and hasn't been merged in v 17 then.
Hopefully, it will been merged in v 17.0.2 or v 18.
